signac/core/buffers.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Buffers for SyncedCollection API."""
+
+import os
+import hashlib
+import logging
+import errno
+from abc import abstractmethod
+
+from .synced_collection import _in_buffered_mode
+from .synced_collection import _get_buffer_force_mode
+from .caching import get_cache
+from .synced_collection import BufferException
+
+logger = logging.getLogger(__name__)
+
+
+class MetadataError(BufferException):
+    """Raised when metadata check fails."""
+
+    def __init__(self, filename):
+        self.filename = filename
+
+    def __str__(self):
+        return f'{self.filename} appears to have been externally modified.'
+
+
+def _hash(blob):
+    """Calculate and return the md5 hash value for the file data."""
+    if blob is not None:
+        m = hashlib.md5()
+        m.update(blob)
+        return m.hexdigest()
+
+
+class FileBuffer:
+    """Implement Buffer for file based backends.
+
+    It provides ``_store_to_buffer`` and ``_flush_buffer`` methods to write the data
+    to the buffer and flush all the data respectively.
+    """
+
+    _cache = get_cache()
+
+    @staticmethod
+    def _get_filemetadata(filename):
+        """Return metadata of file"""
+        try:
+            metadata = os.stat(filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
+                raise
+
+    def _store_to_buffer(self, filename, blob, store_hash=False):
+        """Store the data to the file buffer.
+
+        This method stores the serialized data and metadata of the file to the buffer.
+        It also stores the hash of the data if the ``store_hash`` flag is set.
+        Hash is stored when the data is being loaded from the file and stored in buffer.
+        It is used to skip the extra synchronization (by comparing the hashes).
+        """
+        if store_hash:
+            self._cache['HASH::' + filename] = _hash(blob)
+        if filename not in self._cache:
+            # storing metadata
+            if not (_in_buffered_mode() and _get_buffer_force_mode()):
+                self._cache['METADATA::' + filename] = self._get_filemetadata(filename)
+            # adding filename to list
+            self._cache.setdefault('filenames', [])
+            self._cache['filenames'].append(filename)
+        self._cache[filename] = blob
+
+    @staticmethod
+    @abstractmethod
+    def _write_to_file(filename, blob):
+        """Write data to file."""
+
+    @classmethod
+    def _flush(cls, filename):
+        """Write the data from buffer to the file. Return error if any."""
```

Comment:
Errors are handled by raising now, not returning. This can be removed from the docstring.
```suggestion
        """Write the data from buffer to the file."""
```